pkg/pillar/cmd/zfsmanager/zfsmanager.go

```diff
@@ -0,0 +1,194 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package zfsmanager
+
+import (
+	"flag"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/lf-edge/eve/pkg/pillar/base"
+	"github.com/lf-edge/eve/pkg/pillar/pidfile"
+	"github.com/lf-edge/eve/pkg/pillar/pubsub"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/lf-edge/eve/pkg/pillar/utils"
+	"github.com/lf-edge/eve/pkg/pillar/zfs"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	agentName = "zfsmanager"
+	// Time limits for event loop handlers
+	errorTime            = 3 * time.Minute
+	warningTime          = 40 * time.Second
+	stillRunningInterval = 25 * time.Second
+)
+
+var (
+	debug         bool
+	debugOverride bool
+	logger        *logrus.Logger
+	log           *base.LogObject
+)
+
+type zVolDeviceEvent struct {
+	delete  bool
+	device  string
+	dataset string
+}
+
+type zfsContext struct {
+	zVolStatusPub pubsub.Publication
+}
+
+// Run - an zfs run
+func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject) int {
+	logger = loggerArg
+	log = logArg
+	debugPtr := flag.Bool("d", false, "Debug flag")
+	flag.Parse()
+	debug = *debugPtr
+	debugOverride = debug
+	if debugOverride {
+		logger.SetLevel(logrus.DebugLevel)
+	} else {
+		logger.SetLevel(logrus.InfoLevel)
+	}
+
+	if err := pidfile.CheckAndCreatePidfile(log, agentName); err != nil {
+		log.Fatal(err)
+	}
+
+	// Run a periodic timer so we always update StillRunning
+	stillRunning := time.NewTicker(stillRunningInterval)
+	ps.StillRunning(agentName, warningTime, errorTime)
+
+	ctxPtr := zfsContext{}
+
+	if err := utils.WaitForVault(ps, log, agentName, warningTime, errorTime); err != nil {
+		log.Fatal(err)
+	}
+	log.Functionf("processed Vault Status")
+
+	// Publish cloud metrics
+	zVolStatusPub, err := ps.NewPublication(
+		pubsub.PublicationOptions{
+			AgentName:  agentName,
+			TopicType:  types.ZVolStatus{},
+			Persistent: true,
+		})
+	if err != nil {
+		log.Fatal(err)
+	}
+	ctxPtr.zVolStatusPub = zVolStatusPub
+
+	deviceNotifyChannel := make(chan *zVolDeviceEvent)
+
+	if err := os.MkdirAll(types.ZVolDevicePrefix, os.ModeDir); err != nil {
+		log.Fatal(err)
+	}
+
+	go deviceWatcher(deviceNotifyChannel)
+
+	for {
+		select {
+		case event := <-deviceNotifyChannel:
+			processEvent(ctxPtr, event)
+		case <-stillRunning.C:
+		}
+		ps.StillRunning(agentName, warningTime, errorTime)
+	}
+}
+
+func processEvent(ctxPtr zfsContext, event *zVolDeviceEvent) {
+	if event == nil {
+		return
+	}
+	log.Functionf("processEvent: %+v", event)
+	if event.delete {
+		ctxPtr.zVolStatusPub.Unpublish(event.device)
+		return
+	}
+	ctxPtr.zVolStatusPub.Publish(event.device,
+		types.ZVolStatus{
+			Device:  event.device,
+			Dataset: event.dataset,
+		},
+	)
+}
+
+func deviceWatcher(notifyChannel chan *zVolDeviceEvent) {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Fatalf("NewWatcher: %s", err)
+	}
+	defer w.Close()
+
+	processRecursive := func(dir string) error {
+		// walk through exist files
```

Comment:
existing?

Reply:
But given that the function is called intially plus on each create event, maybe that comment is misleading?

Reply:
It walks the dir subtree.

Reply:
But this is probably not guaranteed to walk/find all files. When you d a w.Add for new directory it will not result in events for any files/dirs under that directory. If you want that guarantee you need a readdir after the w.Add to ensure you see everything.

Reply:
Sure, I call Walk for this case (for observing files inside 'new' directory), because of not support for nested directory notification inside of fsnotify.

Reply:
But the usage is simpler, right? We don't expect nested directories do we?

Reply:
We expect something like `/dev/zvol/persist/volumes/uuid.version`, so, yes, nested directories

Reply:
ok